src/desktop/credentials/DesktopNativeCredentialsFacade.ts

```diff
@@ -1,35 +1,136 @@
 import { CredentialEncryptionMode } from "../../misc/credentials/CredentialEncryptionMode"
-import { DesktopKeyStoreFacade } from "../KeyStoreFacadeImpl"
+import { DesktopKeyStoreFacade } from "../DesktopKeyStoreFacade.js"
 import { DesktopNativeCryptoFacade } from "../DesktopNativeCryptoFacade"
-import { assert } from "@tutao/tutanota-utils"
+import { assert, base64ToUint8Array, uint8ArrayToBase64 } from "@tutao/tutanota-utils"
 import { NativeCredentialsFacade } from "../../native/common/generatedipc/NativeCredentialsFacade.js"
+import { CommonNativeFacade } from "../../native/common/generatedipc/CommonNativeFacade.js"
+import { LanguageViewModel } from "../../misc/LanguageViewModel.js"
+import { DesktopConfig } from "../config/DesktopConfig.js"
+import { DesktopConfigKey } from "../config/ConfigKeys.js"
+import { Argon2idFacade } from "../../api/worker/facades/Argon2idFacade.js"
+import { KEY_LENGTH_BYTES_AES_256 } from "@tutao/tutanota-crypto/dist/encryption/Aes.js"
+import { CryptoError } from "@tutao/tutanota-crypto"
+import { CancelledError } from "../../api/common/error/CancelledError.js"
+import { ProgrammingError } from "../../api/common/error/ProgrammingError.js"
 
+/** the single source of truth for this configuration */
+const SUPPORTED_MODES = [CredentialEncryptionMode.DEVICE_LOCK, CredentialEncryptionMode.APP_PIN] as const
+export type DesktopCredentialsMode = typeof SUPPORTED_MODES[number]
+
+/**
+ *
+ */
 export class DesktopNativeCredentialsFacade implements NativeCredentialsFacade {
-	private readonly _desktopKeyStoreFacade: DesktopKeyStoreFacade
-	private readonly _crypto: DesktopNativeCryptoFacade
+	/**
+	 * @param desktopKeyStoreFacade
+	 * @param crypto
+	 * @param argon2idFacade
+	 * @param lang
+	 * @param conf
+	 * @param getCurrentCommonNativeFacade a "factory" that returns the commonNativeFacade for the window that would be most suited to serve a given request
+	 */
+	constructor(
+		private readonly desktopKeyStoreFacade: DesktopKeyStoreFacade,
+		private readonly crypto: DesktopNativeCryptoFacade,
+		private readonly argon2idFacade: Argon2idFacade,
+		private readonly lang: LanguageViewModel,
+		private readonly conf: DesktopConfig,
+		private readonly getCurrentCommonNativeFacade: () => Promise<CommonNativeFacade>,
+	) {}
 
-	constructor(keyStoreFacade: DesktopKeyStoreFacade, crypto: DesktopNativeCryptoFacade) {
-		this._desktopKeyStoreFacade = keyStoreFacade
-		this._crypto = crypto
+	async decryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		// making extra sure that the mode is the right one since this comes over IPC
+		this.assertSupportedEncryptionMode(encryptionMode)
+		data = await this.maybeDecryptUsingAppPin(data, encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256DecryptKey(credentialsKey, data)
+		return data
```

Comment:
please do not reassign arguments. It is confusing and sometimes deprecated.

```suggestion
		// making extra sure that the mode is the right one since this comes over IPC
		this.assertSupportedEncryptionMode(encryptionMode)
		const dataWithoutPinEncryption = await this.maybeDecryptUsingAppPin(data, encryptionMode)
		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
		return  this.crypto.aes256DecryptKey(credentialsKey, dataWithoutPinEncryption)
```

this is one of the cases where proper naming of each steps actually pays off.